lib/memo_wise/internal_api.rb

```diff
@@ -0,0 +1,244 @@
+    # used to lookup the memoized results specifically for this method and these
+    # parameters.
+    #
+    # @param method_name [Symbol]
+    #   Name of method to derive the "fetch key" for, with given parameters.
+    # @param args [Array]
+    #   Zero or more positional parameters
+    # @param kwargs [Hash]
+    #   Zero or more keyword parameters
+    #
+    # @return [Array, Hash, Object]
+    #   Returns one of:
+    #     - An `Array` if only positional parameters.
+    #     - A nested `Array<Array, Hash>` if *both* positional and keyword.
+    #     - A `Hash` if only keyword parameters.
+    #     - A single object if there is only a single parameter.
+    def fetch_key(method_name, *args, **kwargs) # rubocop:disable Metrics/PerceivedComplexity
+      klass = target.instance_of?(Class) ? target.singleton_class : target.class
+      method = klass.instance_method(method_name)
+
+      if MemoWise::InternalAPI.has_only_required_args?(method)
+        key = method.parameters.map.with_index do |(type, name), index|
+          type == :req ? args[index] : kwargs[name]
+        end
+        key.size == 1 ? key.first : key
+      else
+        has_arg = MemoWise::InternalAPI.has_arg?(method)
+
+        if has_arg && MemoWise::InternalAPI.has_kwarg?(method)
+          [args, kwargs].freeze
+        elsif has_arg
+          args
+        else
+          kwargs
+        end
+      end
+    end
+
+    # Returns visibility of an instance method defined on class `target`.
+    #
+    # @param method_name [Symbol]
+    #   Name of existing *instance* method find the visibility of.
+    #
+    # @return [:private, :protected, :public]
+    #   Visibility of existing instance method of the class.
+    #
+    # @raise ArgumentError
+    #   Raises `ArgumentError` unless `method_name` is a `Symbol` corresponding
+    #   to an existing **instance** method defined on `klass`.
+    #
+    def method_visibility(method_name)
+      if target.private_method_defined?(method_name)
+        :private
+      elsif target.protected_method_defined?(method_name)
+        :protected
+      elsif target.public_method_defined?(method_name)
+        :public
+      else
+        raise ArgumentError,
+              "#{method_name.inspect} must be a method on #{target}"
+      end
+    end
+
+    # Validates that {.memo_wise} has already been called on `method_name`.
+    #
+    # @param method_name [Symbol]
+    #   Name of method to validate has already been setup with {.memo_wise}
+    def validate_memo_wised!(method_name)
+      klass = target.instance_of?(Class) ? target.singleton_class : target.class
+      original_memo_wised_name = :"_memo_wise_original_#{method_name}"
+
+      unless klass.private_method_defined?(original_memo_wised_name)
+        raise ArgumentError, "#{method_name} is not a memo_wised method"
+      end
+    end
+
+    # TODO: Parameter validation for presetting values
```

Comment:
I wonder if we should just remove this placeholder for now? As far as I know, we won't do this for V1, so probably okay to just add it back when we need it?

Reply:
Ok, I'll just go ahead and remove this now then

Reply:
Ok, despite what I wrote in slack, I'll go ahead and just remove this now